libstuff/SSSLState.cpp

```diff
@@ -22,26 +22,45 @@ SSSLState::~SSSLState() {
 }
 
 // --------------------------------------------------------------------------
-SSSLState* SSSLOpen(int s, SX509* x509) {
+SSSLState* SSSLOpen(int s) {
     // Initialize the SSL state
     SASSERT(s >= 0);
     SSSLState* state = new SSSLState;
     state->s = s;
 
-    mbedtls_ctr_drbg_seed(&state->ctr_drbg, mbedtls_entropy_func, &state->ec, 0, 0);
-    mbedtls_ssl_config_defaults(&state->conf, MBEDTLS_SSL_IS_CLIENT, MBEDTLS_SSL_TRANSPORT_STREAM, 0);
+    // All new code here.
+    mbedtls_ssl_init(&state->ssl);
+    mbedtls_ssl_config_init(&state->conf);
+    mbedtls_net_init(&state->net_ctx);
+    state->net_ctx.fd = s;
 
-    mbedtls_ssl_setup(&state->ssl, &state->conf);
+    mbedtls_entropy_init(&state->ec);
+    mbedtls_ctr_drbg_init(&state->ctr_drbg);
+    mbedtls_ctr_drbg_seed(&state->ctr_drbg, mbedtls_entropy_func, &state->ec, nullptr, 0);
+
+    if (mbedtls_ssl_config_defaults(&state->conf, MBEDTLS_SSL_IS_CLIENT, MBEDTLS_SSL_TRANSPORT_STREAM, MBEDTLS_SSL_PRESET_DEFAULT)) {
+        STHROW("ssl config defaults failed");
+    }
 
     mbedtls_ssl_conf_authmode(&state->conf, MBEDTLS_SSL_VERIFY_OPTIONAL);
     mbedtls_ssl_conf_rng(&state->conf, mbedtls_ctr_drbg_random, &state->ctr_drbg);
-    mbedtls_ssl_set_bio(&state->ssl, &state->s, mbedtls_net_send, mbedtls_net_recv, 0);
 
-    if (x509) {
-        // Add the certificate
-        mbedtls_ssl_conf_ca_chain(&state->conf, x509->srvcert.next, 0);
-        SASSERT(mbedtls_ssl_conf_own_cert(&state->conf, &x509->srvcert, &x509->pk) == 0);
+    if (mbedtls_ssl_setup(&state->ssl, &state->conf)) {
+        STHROW("ssl setup failed");
+    }
+
+    /* We don't verify hostnames, and it's also why aboe we set MBEDTLS_SSL_VERIFY_OPTIONAL instead of MBEDTLS_SSL_VERIFY_REQUIRED.
+     * This could be a possible securiy improvement.
+    if (mbedtls_ssl_set_hostname(&state->ssl, "your.server.hostname")) {
+        STHROW("ssl set hostname failed");
```

Comment:
I think we need to set this. If a server requires SNI which is pretty normal these days, the handshake will fail without this set.

Reply:
I mean, that would be great and there's a comment below about it but I think we should do this as a followup because it means we need to pass `hostname` to this function everywhere it's used. Given that we've never done this before, I agree it would be nice to set now, but I things that were working before should still work now without it.